superset/common/query_context_processor.py

```diff
@@ -245,11 +246,20 @@ def normalize_df(self, df: pd.DataFrame, query_object: QueryObject) -> pd.DataFr
             if dttm_col:
                 timestamp_format = dttm_col.python_date_format
 
+        dttm_cols = tuple(
+            label
+            for label in get_base_axis_labels(query_object.columns)
+            if query_object.datasource
+            and (col := query_object.datasource.get_column(label))
+            and col.is_dttm
+        )
+
         normalize_dttm_col(
             df=df,
             timestamp_format=timestamp_format,
             offset=datasource.offset,
             time_shift=query_object.time_shift,
+            dttm_cols=dttm_cols,
```

Comment:
As `timestamp_format` is defined per column, we would probably need to pass a `dict` in that maps the respective column to its `timestamp_format`. So instead of `dttm_cols` being `Optional[Tuple[str, ...]]`, maybe it should be a `Optional[Dict[str, Optional[str]]]` where the key would be the respective `timestamp_format` for said column?

Reply:
Thanks for pointing that out. I am going to refactor `normalize_dttm_col`.